File: modules/drive-sense/android/src/main/java/expo/modules/drivesense/DriveSenseModule.kt
package expo.modules.drivesense

import android.Manifest
import android.content.Context
import android.os.Build
import android.os.PowerManager
import androidx.core.app.ActivityCompat
import expo.modules.interfaces.permissions.PermissionsResponseListener
import expo.modules.interfaces.permissions.PermissionsStatus
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * drive-sense on Android (task N3). README.md in the module root is the binding contract and
 * `src/types.ts` its machine-checked half; `__tests__/native-android.test.ts` holds this file to
 * both (the event list, one `AsyncFunction` per `DRIVE_SENSE_METHODS` entry, the error codes).
 *
 * Rejections are `CodedException`s with the contract codes only (README §2 "Errors"); every other
 * method never rejects.
 */
class DriveSenseModule : Module() {
  companion object {
    /** Every event, for the per-event buffer flush (README §3). Same names as `Events(...)` below. */
    private val OBSERVED_EVENTS = listOf("wake", "activity", "row", "screen", "thermal", "notificationAction", "call")
    private val MODES = setOf("mounted", "pocket", "auto")
  }

  private val context: Context
    get() = appContext.reactContext?.applicationContext ?: throw Exceptions.ReactContextLost()

  private val emitter = EventBus.Emitter { name, body -> sendEvent(name, body) }

  private fun coded(code: String, message: String) = CodedException(code, message, null)

  override fun definition() = ModuleDefinition {
    Name("DriveSense")

    // Keep this list byte-identical to DRIVE_SENSE_EVENTS in src/types.ts and to DriveSenseModule.swift.
    Events("wake", "activity", "row", "screen", "thermal", "notificationAction", "call")

    OnCreate {
      DriveSensePrefs.init(context)
      EventBus.attach(emitter)
    }

    OnDestroy {
      // The JS runtime is going away: nobody listens for rows any more (the watchdog's signal).
      EventBus.detach(emitter)
    }

    for (name in OBSERVED_EVENTS) {
      OnStartObserving(name) { EventBus.startObserving(name) }
      OnStopObserving(name) { EventBus.stopObserving(name) }
    }

    AsyncFunction("arm") { promise: Promise ->
      val ctx = context
      val prefs = DriveSensePrefs.init(ctx)
      if (!DriveSensePermissions.playServicesAvailable(ctx)) {
        prefs.armed = false
        promise.reject(coded("E_UNAVAILABLE", "Google Play services (activity recognition) is not available"))
        return@AsyncFunction
      }
      if (DriveSensePermissions.location(ctx) != "always" || !DriveSensePermissions.motionGranted(ctx)) {
        prefs.armed = false
        ActivityTransitions.unsubscribe(ctx)
        promise.reject(coded("E_PERMISSION", "arm needs location 'always' and motion 'granted'"))
        return@AsyncFunction
      }
      ActivityTransitions.subscribe(ctx) { error ->
        if (error == null) {
          prefs.armed = true
          promise.resolve(null)
        } else {
          prefs.armed = false
          val code = if (error is SecurityException) "E_PERMISSION" else "E_UNAVAILABLE"
          promise.reject(coded(code, "activity transitions could not be requested: ${error.message}"))
        }
      }
    }

    AsyncFunction("disarm") {
      val ctx = context
      ActivityTransitions.unsubscribe(ctx)
      DriveSensePrefs.init(ctx).armed = false
    }

    AsyncFunction("startCapture") { mode: String, promise: Promise ->
      val ctx = context
      if (DriveSensePermissions.location(ctx) == "none") {
        promise.reject(coded("E_PERMISSION", "startCapture needs location permission"))
        return@AsyncFunction
      }
      CaptureService.startFromJs(ctx, if (mode in MODES) mode else "auto") { code ->
        if (code == null) {
          promise.resolve(null)
        } else {
          // From the background with only While-in-use, the refusal is the missing Always (README §2).
          val reason = if (code == "E_FGS_REFUSED" && DriveSensePermissions.location(ctx) == "whenInUse") "E_PERMISSION" else code
          promise.reject(coded(reason, "the location foreground service was refused"))
        }
      }
    }

    AsyncFunction("stopCapture") { promise: Promise ->
      CaptureService.stopFromJs(context) { promise.resolve(null) }
    }

    AsyncFunction("setCaptureRate") { rate: String, promise: Promise ->
      CaptureService.setRate(rate) { promise.resolve(null) }
    }

    AsyncFunction("getState") {
      val ctx = context
      val prefs = DriveSensePrefs.init(ctx)
      val location = DriveSensePermissions.location(ctx)
      val motion = DriveSensePermissions.motion(ctx)
      // `armed` is the effective arming: a revoked permission un-arms (README §2).
      var armed = prefs.armed
      if (armed && (location != "always" || motion != "granted")) {
        prefs.armed = false
        ActivityTransitions.unsubscribe(ctx)
        armed = false
      }
      val capturing = CaptureService.isCapturing
      linkedMapOf<String, Any?>(
        "armed" to armed,
        "capturing" to capturing,
        "rate" to (if (capturing) CaptureService.currentRate ?: "full" else null),
        "mode" to (if (capturing) CaptureService.currentMode ?: prefs.mode else null),
        "platform" to "android",
        "location" to location,
        "motion" to motion,
        "lockSignal" to "reliable",
        "captureWasOpen" to DriveSensePrefs.captureWasOpenAtProcessStart,
        "captureStartedAt" to (if (capturing) CaptureService.captureStartedAt else null),
        "lastRowTs" to CaptureService.lastRowTs
      )
    }

    AsyncFunction("queryMotionHistory") { fromTs: Double, toTs: Double ->
      val ctx = context
      if (DriveSensePermissions.motion(ctx) != "granted") {
        emptyList<Map<String, Any>>()
      } else {
        TransitionStore.query(ctx, fromTs, toTs).map { it.toMap() }
      }
    }

    AsyncFunction("getScreenState") {
      val ctx = context
      linkedMapOf("locked" to PhoneStateReceiver.locked(ctx), "on" to PhoneStateReceiver.screenOn(ctx))
    }

    AsyncFunction("getThermalState") {
      PhoneStateReceiver.thermalLevel(context)
    }

    AsyncFunction("requestMotionPermission") { promise: Promise ->
      val ctx = context
      val prefs = DriveSensePrefs.init(ctx)
      if (!DriveSensePermissions.playServicesAvailable(ctx)) {
        promise.resolve("unavailable")
        return@AsyncFunction
      }
      // Below API 29 activity recognition is an install-time permission.
      if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q || DriveSensePermissions.motionGranted(ctx)) {
        promise.resolve("granted")
        return@AsyncFunction
      }
      val activity = appContext.currentActivity
      val permissions = appContext.permissions
      if (activity == null || permissions == null) {
        promise.resolve("denied") // nothing to prompt from (background)
        return@AsyncFunction
      }
      // "Don't ask again": refused before and the OS will not show the prompt — send to Settings.
      if (prefs.motionRequested &&
        !ActivityCompat.shouldShowRequestPermissionRationale(activity, Manifest.permission.ACTIVITY_RECOGNITION)
      ) {
        promise.resolve("denied")
        return@AsyncFunction
      }
      prefs.motionRequested = true
      val listener = PermissionsResponseListener { result ->
        val status = result[Manifest.permission.ACTIVITY_RECOGNITION]?.status
        promise.resolve(if (status == PermissionsStatus.GRANTED) "granted" else "denied")
      }
      try {
        permissions.askForPermissions(listener, Manifest.permission.ACTIVITY_RECOGNITION)
      } catch (_: Exception) {
        promise.resolve("denied")
      }
    }

    AsyncFunction("excludeFromBackup") { _: String ->
      // Android: backup exclusion is manifest-level (allowBackup false in app.config.ts).
      Unit
    }

    AsyncFunction("setNotificationState") { state: Map<String, Any?> ->
      val stationary = state["stationary"] as? Boolean ?: false
      val startedAt = (state["startedAt"] as? Number)?.toLong()
      CaptureService.setNotificationState(stationary, startedAt)
    }

    AsyncFunction("getLastExitInfo") {
      ExitInfoReader.read(context)
    }

    AsyncFunction("isIgnoringBatteryOptimizations") {
      val ctx = context
      val pm = ctx.getSystemService(Context.POWER_SERVICE) as PowerManager?
      pm?.isIgnoringBatteryOptimizations(ctx.packageName) ?: false
    }

    AsyncFunction("selfTest") { vectorsJson: String ->
      try {
        SelfTest.run(vectorsJson)
      } catch (e: SelfTest.InvalidInput) {
        throw coded("E_INVALID_INPUT", e.message ?: "the vectors JSON cannot be parsed")
      }
    }
  }
}

File: modules/drive-sense/ios/DriveSenseModule.swift
// DriveSense — the RoadWise background drive-capture module (iOS). README.md is the binding
// contract; `src/types.ts` the typed half of it.
//
// The module is a thin bridge: the capture core (`CaptureController`) and the event hub exist
// for the whole process — the app-delegate subscriber creates them at launch, before any module
// does — and every method below runs on the main queue, where they live. Rejections carry only
// the contract's `CodedError` codes.
import ExpoModulesCore

public final class DriveSenseModule: Module, DriveSenseEventSink {
  public func definition() -> ModuleDefinition {
    Name("DriveSense")

    // Keep this list identical to DRIVE_SENSE_EVENTS in src/types.ts and to DriveSenseModule.kt.
    Events("wake", "activity", "row", "screen", "thermal", "notificationAction", "call")

    OnCreate {
      EventHub.shared.attach(self)
    }

    OnDestroy {
      EventHub.shared.detach(self)
    }

    // Buffering (README §3) and the row watchdog (§6) both key on which events have a JS listener.
    OnStartObserving("wake") { EventHub.shared.startObserving("wake") }
    OnStopObserving("wake") { EventHub.shared.stopObserving("wake") }
    OnStartObserving("activity") { EventHub.shared.startObserving("activity") }
    OnStopObserving("activity") { EventHub.shared.stopObserving("activity") }
    OnStartObserving("row") { EventHub.shared.startObserving("row") }
    OnStopObserving("row") { EventHub.shared.stopObserving("row") }
    OnStartObserving("screen") { EventHub.shared.startObserving("screen") }
    OnStopObserving("screen") { EventHub.shared.stopObserving("screen") }
    OnStartObserving("thermal") { EventHub.shared.startObserving("thermal") }
    OnStopObserving("thermal") { EventHub.shared.stopObserving("thermal") }
    OnStartObserving("notificationAction") { EventHub.shared.startObserving("notificationAction") }
    OnStopObserving("notificationAction") { EventHub.shared.stopObserving("notificationAction") }
    OnStartObserving("call") { EventHub.shared.startObserving("call") }
    OnStopObserving("call") { EventHub.shared.stopObserving("call") }

    AsyncFunction("arm") { (promise: Promise) in
      do {
        try CaptureController.shared.arm()
        promise.resolve(nil)
      } catch CaptureController.ArmFailure.unavailable {
        promise.reject("E_UNAVAILABLE", "motion activity is not available on this device")
      } catch CaptureController.ArmFailure.permission(let message) {
        promise.reject("E_PERMISSION", message)
      } catch {
        promise.reject("E_PERMISSION", "arm failed: \(error.localizedDescription)")
      }
    }.runOnQueue(.main)

    AsyncFunction("disarm") { (promise: Promise) in
      CaptureController.shared.disarm()
      promise.resolve(nil)
    }.runOnQueue(.main)

    AsyncFunction("startCapture") { (mode: String, promise: Promise) in
      if CaptureController.shared.startCapture(mode: mode) {
        promise.resolve(nil)
      } else {
        promise.reject("E_PERMISSION", "startCapture needs location permission (whenInUse or always)")
      }
    }.runOnQueue(.main)

    AsyncFunction("stopCapture") { (promise: Promise) in
      CaptureController.shared.stopCapture()
      promise.resolve(nil)
    }.runOnQueue(.main)

    AsyncFunction("setCaptureRate") { (rate: String, promise: Promise) in
      CaptureController.shared.setCaptureRate(rate)
      promise.resolve(nil)
    }.runOnQueue(.main)

    AsyncFunction("getState") { (promise: Promise) in
      promise.resolve(CaptureController.shared.state())
    }.runOnQueue(.main)

    AsyncFunction("queryMotionHistory") { (fromTs: Double, toTs: Double, promise: Promise) in
      ActivitySource.history(fromTs: fromTs, toTs: toTs) { promise.resolve($0) }
    }.runOnQueue(.main)

    AsyncFunction("getScreenState") { (promise: Promise) in
      let s = PhoneState.snapshot()
      promise.resolve(["locked": s.locked, "on": s.screenOn])
    }.runOnQueue(.main)

    AsyncFunction("getThermalState") { (promise: Promise) in
      promise.resolve(PhoneState.thermalLevel())
    }.runOnQueue(.main)

    AsyncFunction("requestMotionPermission") { (promise: Promise) in
      ActivitySource.requestPermission { result in
        CaptureController.shared.enforceArmingPermissions()
        promise.resolve(result)
      }
    }.runOnQueue(.main)

    AsyncFunction("excludeFromBackup") { (uri: String, promise: Promise) in
      do {
        try Backup.exclude(uri)
        promise.resolve(nil)
      } catch Backup.Failure.notFound(let message) {
        promise.reject("E_NOT_FOUND", message)
      } catch Backup.Failure.failed(let message) {
        promise.reject("E_IO", message)
      } catch {
        promise.reject("E_IO", error.localizedDescription)
      }
    }

    // Android S3's notification; iOS has none.
    AsyncFunction("setNotificationState") { (_: NotificationStateRecord, promise: Promise) in
      promise.resolve(nil)
    }

    // Android's ApplicationExitInfo has no iOS equivalent.
    AsyncFunction("getLastExitInfo") { (promise: Promise) in
      promise.resolve(nil)
    }

    // iOS has no per-app battery restriction for the user to lift (README §2).
    AsyncFunction("isIgnoringBatteryOptimizations") { (promise: Promise) in
      promise.resolve(true)
    }

    // Heavy: runs on the module's own async queue, never the main queue.
    AsyncFunction("selfTest") { (vectorsJson: String, promise: Promise) in
      do {
        promise.resolve(try SelfTest.run(vectorsJson))
      } catch let e as SelfTest.InvalidInput {
        promise.reject("E_INVALID_INPUT", e.message)
      } catch {
        promise.reject("E_INVALID_INPUT", error.localizedDescription)
      }
    }
  }

  // MARK: - DriveSenseEventSink

  func deliver(_ event: String, _ body: [String: Any]) {
    sendEvent(event, body.mapValues { $0 as Any? })
  }
}

/// `setNotificationState`'s argument (ignored on iOS; declared so the bridge accepts it).
struct NotificationStateRecord: Record {
  @Field var stationary: Bool = false
  @Field var startedAt: Double?
}
